src/wallet/wallet2.cpp

```diff
@@ -662,44 +732,72 @@ void wallet2::process_new_transaction(const cryptonote::transaction& tx, const s
 	m_callback->on_skip_transaction(height, tx);
       return;
     }
+    tx_pub_key = pub_key_field.pub_key;
+
+    // We got a payment ID to go with this tx
+    if (pID_info.has_payment_id8) {
+      LOG_PRINT_L2("Found encrypted payment ID: " << pID_info.payment_id8);
+      decrypted_payment_id8 = pID_info.payment_id8;
+      if (!decrypt_payment_id(decrypted_payment_id8, tx_pub_key, m_account.get_keys().m_view_secret_key))
+      {
+        LOG_PRINT_L0("Failed to decrypt payment ID: " << decrypted_payment_id8);
+      }
+      else
+      {
+        LOG_PRINT_L2("Decrypted payment ID: " << decrypted_payment_id8);
+        // put the 64 bit decrypted payment id in the first 8 bytes
+        memcpy(decrypted_payment_id32.data, decrypted_payment_id8.data, 8);
+        // rest is already 0, but guard against code changes above
+        memset(decrypted_payment_id32.data + 8, 0, 24);
+      }
+    }
 
     int num_vouts_received = 0;
-    tx_pub_key = pub_key_field.pub_key;
     bool r = true;
     std::deque<cryptonote::keypair> in_ephemeral(tx.vout.size());
     std::deque<crypto::key_image> ki(tx.vout.size());
     std::deque<uint64_t> amount(tx.vout.size());
     std::deque<rct::key> mask(tx.vout.size());
     int threads = tools::get_max_concurrency();
     const cryptonote::account_keys& keys = m_account.get_keys();
-    crypto::key_derivation derivation;
-    generate_key_derivation(tx_pub_key, keys.m_view_secret_key, derivation);
+    std::pair<crypto::secret_key, crypto::public_key> derivation_info;
+    derivation_info.first  = keys.m_view_secret_key;
+    derivation_info.second = tx_pub_key;
     if (miner_tx && m_refresh_type == RefreshNoCoinbase)
     {
       // assume coinbase isn't for us
     }
     else if (miner_tx && m_refresh_type == RefreshOptimizeCoinbase)
     {
       uint64_t money_transfered = 0;
-      bool error = false, received = false;
-      check_acc_out_precomp(keys.m_account_address.m_spend_public_key, tx.vout[0], derivation, 0, received, money_transfered, error);
+      bool error = false;
+      is_received_info received;
+      check_acc_out_precomp_onetime(keys.m_account_address.m_spend_public_key, tx.vout[0], derivation_info, pID_info.onetime_c, 0, received, money_transfered, error);
       if (error)
       {
         r = false;
       }
       else
       {
         // this assumes that the miner tx pays a single address
-        if (received)
+        if (received.standard || received.onetime)
         {
-          wallet_generate_key_image_helper(keys, tx_pub_key, 0, in_ephemeral[0], ki[0]);
+          is_onetime = is_onetime || received.onetime;
```

Comment:
It is not illegal at the protocol level. There is just code to validate a transfer's parameters before it is made, and while it might already prevent sending to a one time address and using a payment id, that's what I am trying to say needs to be checked.